x/capability/keeper/keeper.go

```diff
@@ -0,0 +1,244 @@
+	}
+}
+
+// ScopeToModule attempts to create and return a ScopedKeeper for a given module
+// by name. It will panic if the keeper is already sealed or if the module name
+// already has a ScopedKeeper.
+func (k *Keeper) ScopeToModule(moduleName string) ScopedKeeper {
+	if k.sealed {
+		panic("cannot scope to module via a sealed capability keeper")
+	}
+
+	if _, ok := k.scopedModules[moduleName]; ok {
+		panic(fmt.Sprintf("cannot create multiple scoped keepers for the same module name: %s", moduleName))
+	}
+
+	k.scopedModules[moduleName] = struct{}{}
+
+	return ScopedKeeper{
+		cdc:      k.cdc,
+		storeKey: k.storeKey,
+		memKey:   k.memKey,
+		module:   moduleName,
+	}
+}
+
+// InitializeAndSeal loads all capabilities from the persistent KVStore into the
+// in-memory store and seals the keeper to prevent further modules from creating
+// a scoped keeper. InitializeAndSeal must be called once after the application
+// state is loaded.
+func (k *Keeper) InitializeAndSeal(ctx sdk.Context) {
+	if k.sealed {
+		panic("cannot initialize and seal an already sealed capability keeper")
+	}
+
+	memStore := ctx.KVStore(k.memKey)
+	memStoreType := memStore.GetStoreType()
+
+	if memStoreType != sdk.StoreTypeMemory {
+		panic(fmt.Sprintf("invalid memory store type; got %s, expected: %s", memStoreType, sdk.StoreTypeMemory))
+	}
+
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixIndexCapability)
+	iterator := sdk.KVStorePrefixIterator(prefixStore, nil)
+
+	// initialize the in-memory store for all persisted capabilities
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		index := types.IndexFromKey(iterator.Key())
+		cap := types.NewCapabilityKey(index)
+
+		var capOwners *types.CapabilityOwners
+		k.cdc.MustUnmarshalBinaryBare(iterator.Value(), &capOwners)
+
+		for _, owner := range capOwners.Owners {
+			// Set the forward mapping between the module and capability tuple and the
+			// capability name in the in-memory store.
+			memStore.Set(types.FwdCapabilityKey(owner.Module, cap), []byte(owner.Name))
+
+			// Set the reverse mapping between the module and capability name and the
+			// capability in the in-memory store.
+			memStore.Set(types.RevCapabilityKey(owner.Module, owner.Name), k.cdc.MustMarshalBinaryBare(cap))
+		}
+	}
+
+	k.sealed = true
+}
+
+// NewCapability attempts to create a new capability with a given name. If the
+// capability already exists in the in-memory store, an error will be returned.
+// Otherwise, a new capability is created with the current global index. The
+// newly created capability has the module and name tuple set as the initial owner.
+// Finally, the global index is incremented along with forward and reverse indexes
+// set in the in-memory store.
+func (sk ScopedKeeper) NewCapability(ctx sdk.Context, name string) (types.Capability, error) {
+	store := ctx.KVStore(sk.storeKey)
+	memStore := ctx.KVStore(sk.memKey)
+
+	if memStore.Has(types.RevCapabilityKey(sk.module, name)) {
```

Comment:
No, the namespacing is completely local, which is safe since it is prefixed with `sk.module`, and no two `ScopedKeeper`s can have the same module name.

Reply:
(the `index` is the part that is globally unique)

Reply:
Updated `NewCapability` to reflect this.